autokeras/utils.py

```diff
@@ -0,0 +1,100 @@
+import numpy as np
+import tensorflow as tf
+from tensorflow.python.util import nest
+
+
+def get_global_average_pooling_layer(shape):
+    return [tf.keras.layers.GlobalAveragePooling1D,
+            tf.keras.layers.GlobalAveragePooling2D,
+            tf.keras.layers.GlobalAveragePooling3D][len(shape) - 3]
+
+
+def get_global_max_pooling_layer(shape):
+    return [tf.keras.layers.GlobalMaxPool1D,
+            tf.keras.layers.GlobalMaxPool2D,
+            tf.keras.layers.GlobalMaxPool3D][len(shape) - 3]
+
+
+def format_inputs(inputs, name=None, num=None):
+    inputs = nest.flatten(inputs)
+    if not isinstance(inputs, list):
+        inputs = [inputs]
+
+    if num is None:
+        return inputs
+
+    if not len(inputs) == num:
+        raise ValueError('Expected {num} elements in the '
+                         'inputs list for {name} '
+                         'but received {len} inputs.'.format(num=num,
+                                                             name=name,
+                                                             len=len(inputs)))
+    return inputs
+
+
+def split_train_to_valid(x, y, validation_split):
+    # Generate split index
+    validation_set_size = int(len(x[0]) * validation_split)
+    validation_set_size = max(validation_set_size, 1)
+    validation_set_size = min(validation_set_size, len(x[0]) - 1)
+
+    # Split the data
+    x_train = []
+    y_train = []
+    x_val = []
+    y_val = []
+    for temp_x in x:
+        x_train.append(temp_x[:-validation_set_size])
+        x_val.append(temp_x[-validation_set_size:])
+    for temp_y in y:
+        y_train.append(temp_y[:-validation_set_size])
+        y_val.append(temp_y[-validation_set_size:])
+
+    return (x_train, y_train), (x_val, y_val)
+
+
+def get_name_scope():
+    with tf.name_scope('a') as scope:
+        name_scope = scope[:-2]
+    return name_scope
+
+
+def dataset_shape(dataset):
+    return tf.compat.v1.data.get_output_shapes(dataset)
+
+
+def input_list_to_dataset(x):
+    new_x = []
+    for temp_x in x:
+        if isinstance(temp_x, np.ndarray):
+            new_x.append(tf.data.Dataset.from_tensor_slices(temp_x))
+    return new_x
+
+
+def prepare_preprocess(x, y=None, validation_data=None):
+    """Convert each input to a tf.data.Dataset."""
+    x = format_inputs(x, 'train_x')
+    x = input_list_to_dataset(x)
+    if y:
+        y = format_inputs(y, 'train_y')
+        y = input_list_to_dataset(y)
+    if validation_data:
+        x_val, y_val = validation_data
+        x_val = format_inputs(x_val, 'x_val')
+        y_val = format_inputs(y_val, 'y_val')
+        x_val = input_list_to_dataset(x_val)
+        y_val = input_list_to_dataset(y_val)
+        validation_data = x_val, y_val
+    return x, y, validation_data
```

Comment:
This returns a tuple of lists of datasets, but really it should return 2 datasets, one for the training data (that yields x, y) and one for the validation data (that yields x_val, y_val).

Reply:
If the inputs are all patched into one Dataset, they cannot be passed to different preprocessing blocks separately. Each preprocessor would get all the inputs, they have to know which one to use.